[SPARK-46366][SQL] Use WITH expression in BETWEEN to avoid duplicate expressions

dbatomic opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'core/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
Prior to this change `e BETWEEN lower AND upper` expression used to be transformed into `lower <= e && e <= upper`. This means that `e` would be evaluated twice which is problematic from both correctness and performance perspectives.

Suggested fix is to use `WITH` expression that was introduced with [this](https://github.com/apache/spark/commit/01c294b05f3a9b7bd87cda0ee8b0160f5f58bb24) change.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
Current implementation is not correct for non deterministic expressions, since two calls might return different results.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
With this change generated plan for BETWEEN statement will be different. An example of generated plan is provided in tests.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->
Existing tests plus new test in PlanGenerationTestSuite.

### Was this patch authored or co-authored using generative AI tooling?
<!--
If generative AI tooling has been used in the process of authoring this patch, please include the
phrase: 'Generated-by: ' followed by the name of the tool and its version.
If no, write 'No'.
Please refer to the [ASF Generative Tooling Guidance](https://www.apache.org/legal/generative-tooling.html) for details.
-->
Yes.

dbatomic committed:
intial prototype

dbatomic committed:
Adding test to plan generation test suite

dbatomic committed:
Adding explain check

dbatomic committed:
adding new comment

dbatomic committed:
Merge remote-tracking branch 'upstream/master' into between_expression_v2

dbatomic committed:
Merge branch 'apache:master' into between_expression_v2

dbatomic committed:
fixing test in expressionparsersuite

dbatomic committed:
Fixing streaming logic

dbatomic committed:
Updating golden files

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/betweenExpression.scala:
```diff
@@ -0,0 +1,49 @@
+  """,
+  since = "2.0.0",
+  group = "conditional_funcs")
+case class BetweenExpr(proj: Expression, lower: Expression,
+                       upper: Expression, replacement: Expression)
```
Please, adjust indentations according to https://github.com/databricks/scala-style-guide?tab=readme-ov-file#indent

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/betweenExpression.scala:
```diff
@@ -0,0 +1,49 @@
+
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = "BETWEEN(projection, lower, upper) - Returns true if `projection` > `lower` and `projection` <  `upper`.",
```
Could you describe arguments too in `arguments`

dtenedor commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:
```diff
@@ -3436,6 +3437,21 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
+  /**
+   * Transform UnresolvedBetweenExpression into a [BetweenExpr].
+   */
+  object ResolveBetweenExpression extends Rule[LogicalPlan] {
```
We should avoid trying to add new analyzer rules at this point. Could we accomplish this  by combining `UnresolvedBetweenExpression` and `ResolvedBetweeExpression` into just one `BetweenExpression` and make that inherit from `RuntimeReplaceable`? Then its `replacement` can be the `With` expression of interest.

![image](https://github.com/apache/spark/assets/99207096/f96c15a0-bd5a-4870-b13f-2150ff2e58ff)

dbatomic replied:
This is what I tried initially. I tried to explain the problem in PR description, but I guess that I wasn't clear enough. Let me try to do this again:
Initially I tried with `BetweenExpr extends RuntimeReplaceable`. The problem that I hit was that `CommonExpressionRef` required dataType of `CommonExpressionDef` to be known for it to be created:
```scala
  def this(exprDef: CommonExpressionDef) = this(exprDef.id, exprDef.dataType, exprDef.nullable)
```
To get around this I would like to create replacement of BetweenExpr after I resolve it's Child expressions and hence after I have it's dataType. I hit a wall doing this so I did this thing with UnresolvedBetween->BetweenExpr (for which I will know the types)->With, which works, but I agree that it would be better if we could avoid this + get rid of extra rule.

I will give it another try. If you guys know a way to get around this I would appreciate some help :)

dbatomic replied:
Ok, I hope that i got it right in the latest iteration. Please take a look.

Btw, thanks for the comment, this is definitely much cleaner.

dbatomic committed:
Removing unresolved between expression.

dbatomic committed:
Renaming BetweenExpr -> Between

dbatomic committed:
Fixing golden files

dbatomic committed:
Adding comments.

dbatomic committed:
Adding more rigid coercion rules for between, there are more alike if/else combination.
Even with this rule there are changes in DS suite. We should work to eliminate these.

dbatomic committed:
Using built in function for resolution.

dbatomic committed:
Updating changed plans

dbatomic committed:
Updating golden files

dtenedor commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala:
```diff
@@ -775,6 +775,7 @@ object FunctionRegistry {
     expression[In]("in"),
     expression[Not]("not"),
     expression[Or]("or"),
+    expression[Between]("between"),
```
please sort alphabetically (move to line 775)?

dtenedor reviewed (approved):
LGTM!

dbatomic committed:
Updating golden files for sql-expression-schema. Removing between from operator list since it is now in function list.

dbatomic committed:
Doc update for between expression

dbatomic committed:
Updating expression documentation

dbatomic committed:
Merge branch 'apache:master' into between_expression_v2

dbatomic committed:
fixing failing doc test

dbatomic committed:
Documentation update

dbatomic committed:
Merge branch 'apache:master' into between_expression_v2

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/StreamingJoinHelper.scala:
```diff
@@ -96,17 +96,22 @@ object StreamingJoinHelper extends PredicateHelper with Logging {
-        logInfo(s"Condition $joinCondition generated watermark constraint = ${stateWatermark.get}")
-      }
-      stateWatermark
+      val stateWatermarks = (predicate match {
```
why do we change the streaming code?

dbatomic replied:
I got failures in streaming tests. It turns out that they calculate upper/lower bounds based on `if x > y` conditions. With `Between` op we no longer have expressions they can use so I had to extend the logic to cover `Between` for this watermark resolution.

cloud-fan replied:
I think we can keep the code structure unchanged, with one more patter match
```
case Between(p, lower, upper, _) =>
  getStateWatermarkSafely(lower, p).map(_ - 1).orElse(getStateWatermarkSafely(p, upper).map(_ - 1))
```

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala:
```diff
@@ -479,10 +479,8 @@ abstract class TypeCoercionBase {
             Coalesce(es.map(castIfNotSameType(_, finalDataType)))
-          case None =>
-            c
+          case None => c
```
let's avoid unnecessary changes.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Between.scala:
```diff
@@ -0,0 +1,60 @@
+  """,
+  since = "4.0.0",
+  group = "conditional_funcs")
+case class Between(proj: Expression, lower: Expression, upper: Expression, replacement: Expression)
```
```suggestion
case class Between private(proj: Expression, lower: Expression, upper: Expression, replacement: Expression)
```

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Between.scala:
```diff
@@ -0,0 +1,60 @@
+    this(proj, lower, upper, {
+        val commonExpr = CommonExpressionDef(proj)
+        val ref = new CommonExpressionRef(commonExpr)
+        With(And(
```
to make the code a bit more clearer:
```
val replacement = And(GreaterThanOrEqual(ref, lower), LessThanOrEqual(ref, upper))
With(replacement, Seq(commonExpr))
```

cloud-fan commented on connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/PlanGenerationTestSuite.scala:
```diff
@@ -516,6 +516,10 @@ class PlanGenerationTestSuite
   }
 
+  test("between expr") {
+    simple.selectExpr("rand(123) BETWEEN 0.1 AND 0.2")
```
can we add this test to a golden file test as well? This one is only for Spark Connect, we need to test the classic Spark SQL too.

cloud-fan replied:
I mean this golden file: sql/core/src/test/resources/sql-tests/inputs/predicate-functions.sql

cloud-fan replied:
ping @dbatomic , this is not addressed yet.

dbatomic committed:
Resolving PR comments

dbatomic committed:
Merge branch 'between_expression_v2' of github.com:dbatomic/spark into between_expression_v2
merge

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Between.scala:
```diff
@@ -0,0 +1,58 @@
+  """,
+  arguments = """
+    Arguments:
+      * projection - An expression that is being compared with lower and upper bound.
```
The parameter name `projection` is a bit weird. How about just `value` or `input`?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Between.scala:
```diff
@@ -0,0 +1,58 @@
+  """,
+  since = "4.0.0",
+  group = "conditional_funcs")
+case class Between private(proj: Expression, lower: Expression, upper: Expression, replacement: Expression)
```
ditto, `value` or `input`

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Between.scala:
```diff
@@ -0,0 +1,58 @@
+  group = "conditional_funcs")
+case class Between private(proj: Expression, lower: Expression, upper: Expression, replacement: Expression)
+  extends RuntimeReplaceable with InheritAnalysisRules  {
+  def this(proj: Expression, lower: Expression, upper: Expression) = {
```
ditto

dbatomic committed:
PR comments

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Between.scala:
```diff
@@ -0,0 +1,58 @@
+
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = "Usage: expr1 [NOT] BETWEEN expr2 AND expr3 - evaluate if `expr1` is [not] in between `expr2` and `expr3`",
```
let's use the parameter names in the usage doc as well.

cloud-fan commented:
let's not forget about this comment :)  https://github.com/apache/spark/pull/44299/files#r1436957948

dbatomic committed:
PR comment on better documentation

dbatomic committed:
doc update

dbatomic committed:
adding rand example to predicate function golden files.

cloud-fan commented:
thanks, merging to master!
